waspc/data/Generator/templates/react-app/src/webSocket.ts

```diff
@@ -0,0 +1,71 @@
+{{={= =}=}}
+
+import { useState, useEffect } from 'react'
+import { io, Socket } from 'socket.io-client'
+import { DefaultEventsMap, EventsMap } from '@socket.io/component-emitter'
+
+import { getAuthToken, events as apiEvents } from './api'
+import config from './config'
+
+// TODO: In the future, it would be nice if users could pass more
+// options to `io`, likely via some `configFn`.
+export const socket = io(config.apiUrl, { autoConnect: {= autoConnect =} })
+
+function refreshAuthToken() {  
+  // NOTE: In the future, we should consider making this explicit in the Wasp file when
+  // we make the change for how auth works in Operations.
+  // For now, it is fine, and can be ignored on the server if not needed.
+  socket.auth = {
+    token: getAuthToken()
+  }
+
+  if (socket.connected) {
+    socket.disconnect()
+    socket.connect()
+  }
+}
+
+refreshAuthToken()
+apiEvents.on('authToken.set', refreshAuthToken)
+apiEvents.on('authToken.clear', refreshAuthToken)
+
+export function useSocket<
+  ServerToClientEvents extends EventsMap = DefaultEventsMap,
+  ClientToServerEvents extends EventsMap = DefaultEventsMap
+>(): {
+  socket: Socket<ServerToClientEvents, ClientToServerEvents>;
+  isConnected: boolean;
+  registerHandler: typeof registerHandler;
+} {
+  const [isConnected, setIsConnected] = useState(socket.connected)
+
+  useEffect(() => {
+    function onConnect() {
+      setIsConnected(true)
+    }
+
+    function onDisconnect() {
+      setIsConnected(false)
+    }
+
+    socket.on('connect', onConnect)
+    socket.on('disconnect', onDisconnect)
+
+    return () => {
+      socket.off('connect', onConnect)
+      socket.off('disconnect', onDisconnect)
+    }
+  }, [])
+
+  function registerHandler<Event extends keyof ServerToClientEvents>
+    (event: Extract<Event, string>, cb: ServerToClientEvents[Event]) {
+    useEffect(() => {
+      socket.on(event, cb)
+      return () => {
+        socket.off(event, cb)
+      }
+    }, [])
+  }  
```

Comment:
Is this valid regarding the rules that hooks must obey? I thought that hook needs to be used in component at the top level, or in another hook. No branching.

What we do is have a hook that returns a function which then calls useEffect (another hook) inside of it, I find that a bit weird and potentially breaking those rules. They are calling a hook, but it looks like it is not a hook, ... .

> useEffect is a Hook, so you can only call it at the top level of your component or your own Hooks. You can’t call it inside loops or conditions. If you need that, extract a new component and move the state into it.

Maybe it passes React checks somehow, but it doesn't look to me like it clearly obeys these rules.

To me it sounds like the better way would be that we have a separate hook called `useSocketListener` that can be used to register a listener.

Reply:
I just showed this code to chatGPT + I gave it that piece of react docs and it suggested on its own that we create another hook, like I suggested above. I am not sure if that is much of evidence because I kind of pushed it into that line of thinking and chatGPT is very agreeable but still :D.

Reply:
Yeah I agree, I felt funny doing it this way too, but couldn't see any other way to do what @infomiho suggested and it seemed to work fine. Conceptually I think it is probably "safe"/valid, but perhaps for safety, a separate hook is preferable though. I'll extract 👍🏻

Reply:
I also don't see a better way to do what @infomiho suggested, but on the other hand the idea of `useSocketListener` hook sounds good to me!

Reply:
He was going to take a look to make sure I was doing it right, but thought it was valid. If so, maybe we can just leave as is?

Reply:
I have to admit I don't see how it could be valid with that approach. But I am not a React expert hm.

I would just go for the separate hook. But if you prefer waiting for @infomiho to give his take on this, we can also do that!

Reply:
Ok, after I extracted it, I can kinda see how these should be equivalent now. In the former case, we are just providing back a reference to a function (a custom hook in both cases), not really different from it being at the top level. The downside with a separate hook is it's more verbose since it requires adding the type:

<img width="711" alt="Screenshot 2023-06-01 at 1 12 49 PM" src="https://github.com/wasp-lang/wasp/assets/523636/c00eb081-e507-471d-8ac6-04eb64c75819">
<img width="838" alt="Screenshot 2023-06-01 at 1 13 02 PM" src="https://github.com/wasp-lang/wasp/assets/523636/9a3edd17-b651-4e96-ad36-ae087befb8c7">

Reply:
Hm too bad that we have to repeat that type but I guess we can't avoid it.

I am ok also with the first option in general, the problem is that I am suspicious it breaks the rules what React requires from us for using hooks, so I wonder if it will misbehave down the road possibly. I could ask in our Discord.

Reply:
Ok did some more research:

Rules from React docs:

> Only Call Hooks at the Top Level
> Don’t call Hooks inside loops, conditions, or nested functions. Instead, always use Hooks at the top level of your React function, before any early returns. By following this rule, you ensure that Hooks are called in the same order each time a component renders. That’s what allows React to correctly preserve the state of Hooks between multiple useState and useEffect calls. (If you’re curious, we’ll explain this in depth [below](https://legacy.reactjs.org/docs/hooks-rules.html#explanation).)
> 
> Only Call Hooks from React Functions
> Don’t call Hooks from regular JavaScript functions. Instead, you can:
> 
> ✅ Call Hooks from React function components.
> ✅ Call Hooks from custom Hooks (we’ll learn about them [on the next page](https://legacy.reactjs.org/docs/hooks-custom.html)).
> By following this rule, you ensure that all stateful logic in a component is clearly visible from its source code.

By having this extra step where we return a hook from another function (although it is hook), I think we violate these rules.

@Zeko369 also showed on Discord that this can break easily actually, so that confirms it: https://discord.com/channels/686873244791210014/1080864617347162122/1114104310645669928 .

Reply:
Start of Discord convo: https://discord.com/channels/686873244791210014/1080864617347162122/1113918843774054440 .

Reply:
Pulled up to top level 👍🏻

Reply:
The only thing I am curious about now is, if we actually return it from a useSocket hook -> so if useSocket hook returns useSocketListener hook, which is weird, but -> does that mean that useSocketListener hook has some extra powers, like has the types inferred? Although there is also that idea that I mentioned above, that all the types could be already provided by Wasp in the "backstage" if we registered them with Wasp, since we only have one socket and we actually know the types for it. Hmmmmmmmmmmm.

Reply:
If we provided types to Wasp, in Wasp DSL, via ext imports, then Wasp could, in generated Wasp code, specify that these functions have those types!

Reply:
Would that be kinda clunky from a DX perspective to have 4 imports (presumably so we can differentiate them regardless of type name)?

Reply:
Maybe. They could maybe also be provided via JS, via some kind of "hook" (not React hook) and then you import (in Wasp) just that one thing, if we feel bad about having too many ext imports in Wasp. But it improves DX of `useSocket` potentially, because you don't have to provide types there.
On the other hand, they could create their own hook, useSocketsTyped which just calls useSocket with types and they can then use that everywhere. Hm.